### ISSUE TITLE
Synchronized Batch normalization in multi GPUs training

### ISSUE THREAD

**Issue by cccorn** [apache/incubator-mxnet#8458]
I found that Batch Norm dose not synchronize the mean and variance through multi-GPUs in training phase.But I want to train ResNet using big image, so i can only use a small batch in each device. I think it is not appropriate to use Batch Norm when the batch size is too small (5 images).
Is there any way to easily solve this problem?

**Comment by Tommymhz** [comment]
Also need the Synchronized Batch normalization!!!!1

**Comment by Tommymhz** [comment]
@tqchen @pluskid  @mli

**Comment by Jerryzcn** [comment]
Actually all major framework does not synchronize mean and variance, since this will significantly slow down the training, if the cross bandwidth is low. However, i think this is a very useful feature. For semantic segmentation, especially reproducing PSPNet, synchronized batch norm is necessary.

**Comment by Tommymhz** [comment]
So, How to implement it ?! @Jerryzcn  it seem that it's import for segmentation

**Comment by Jerryzcn** [comment]
@Tommymhz  I would suggest implement your own batchnorm in gluon, and sync across all gpus.

**Comment by Tommymhz** [comment]
However, My code are based on the symbol !!

**Comment by coolbrain** [comment]
I need the Synchronized Batch normalization, too. My code is based on symbol. Any Suggestion?

**Comment by KeyKy** [comment]
Could you share your code?

**Comment by Jerryzcn** [comment]
@zhanghang1989

**Comment by zhanghang1989** [comment]
There is no easy way to solve Sync BN. I have implemented it using PyTorch (http://hangzh.com/PyTorch-Encoding/notes/syncbn.html) by rewriting DataParallel and cross-GPU communication. I will implement MXNet Gluon version as well, probably release it early next year.

**Comment by John1231983** [comment]
@zhanghang1989 : how about tensorflow. Could you also implement it? We are really need it in tensorflow

**Comment by marcoabreu** [comment]
@John1231983 This repository is for MXNet and not tensor flow. Please refrain from requests regarding other frameworks.

**Comment by oronanschel** [comment]
Hi, did someone implemented that yet?

**Comment by zhanghang1989** [comment]
MXNet Gluon Cross-GPU Batch Norm is Implemented here https://github.com/zhanghang1989/MXNet-Gluon-SyncBN . Feel free to try and leave comments.

**Comment by cccorn** [comment]
@zhanghang1989 Thanks!

**Comment by thomelane** [comment]
@sandeep-krishnamurthy this issue can be closed now, thanks. further discussions should be on the PR https://github.com/apache/incubator-mxnet/pull/11502

**Comment by pengwangucla** [comment]
> There is no easy way to solve Sync BN. I have implemented it using PyTorch (http://hangzh.com/PyTorch-Encoding/notes/syncbn.html) by rewriting DataParallel and cross-GPU communication. I will implement MXNet Gluon version as well, probably release it early next year.

HI Hang, I used your sync_bn implementation for mxnet symbol. However, it reduced the performance of my network. I wonder whether you have ever tried with symbol API with your sync_bn other than gluon. Thanks

**Comment by zhanghang1989** [comment]
The SyncBN does reduce the speed due to synchronization for every BN layer during both forward and backward. Use it only when it is necessary, especially on computation intensive task such as semantic segmentation or object detection where the synchronization is not a huge bottleneck comparing to the computation.

**Comment by pengwangucla** [comment]
I understand it reduces speed, but it also reduces my accuracy, which makes me really confused.

> The SyncBN does reduce the speed due to synchronization for every BN layer during both forward and backward. Use it only when it is necessary, especially on computation intensive task such as semantic segmentation or object detection where the synchronization is not a huge bottleneck comparing to the computation.

Thanks for the reply, I understand it reduces the speed as I observed, but it also reduces the accuracy from a pretrained se-net. I wonder whether you have tried it using symbol API ?  Thanks
